<?xml version="1.0"?>
<launch>
  <arg name="device"    default="/dev/video0"/>
  <arg name="width"     default="640"/>
  <arg name="height"    default="480"/>
  <arg name="framerate" default="15"/>

  <node pkg="usb_cam" type="usb_cam_node" name="usb_cam" output="screen">
    <param name="video_device"    value="$(arg device)"/>
    <param name="image_width"     value="$(arg width)"/>
    <param name="image_height"    value="$(arg height)"/>
    <param name="framerate"       value="$(arg framerate)"/>
    <param name="pixel_format"    value="yuyv"/>
    <param name="camera_frame_id" value="usb_cam"/>
    <param name="io_method"       value="mmap"/>

    <remap from="/usb_cam/image_raw" to="/image_in"/>
  </node>

  <node name="dynamic_reconfigure_load" pkg="dynamic_reconfigure" type="dynparam" args="load color_extraction_params $(find opengl_ros)/params/color_extraction.yaml" />
  <node pkg="opengl_ros" type="simple_renderer"
    name="simple_renderer" output="screen">
    <param name="width"           value="$(arg width)"/>
    <param name="height"          value="$(arg height)"/>
    <param name="vertex_shader"   value="$(find opengl_ros_lib)/shader/vs_passthrough.glsl"/>
    <param name="fragment_shader" value="$(find opengl_ros)/shader/fs_color_extraction.glsl"/>
    <param name="image_in"       value="/image_in"/>
    <param name="image_out"      value="/image_out"/>
  </node>

  <node pkg="rqt_reconfigure" type="rqt_reconfigure" name="reconfigure_gui"/> 

  <node pkg="image_view" type="image_view" name="viewer_in">
    <remap from="image" to="/image_in"/>
    <param name="window_name" value="Input Image"/>
  </node>

  <node pkg="image_view" type="image_view" name="viewer_out">
    <remap from="image" to="/image_out"/>
    <param name="window_name" value="Output Image"/>
  </node>
</launch>
